ebpf_program_t computes hash of program info on first attach

Alan-Jowett opened the pull request:
Signed-off-by: Alan Jowett <alanjo@microsoft.com>

Resolves: #1910 

## Description

Verification of JIT and Interpreted programs is as follows:
1) User program create BPF program and map objects.
2) User program passes handle to program, handle to maps, and byte code to eBPF Service.
3) eBPF Service queries program information from the handle and verifies byte code against program information returned by that handle.
4) eBPF Service loads the bytecode or machine code into the program object.
5) Program is then attached and invoked.

To resolve cases where program information changes between 1 and 5, runtime captures a SHA256 hash of the program info when the program is initially created (step 1) and then recomputes if the provider detaches and reattaches. If it doesn't match, it refuses the attach.

## Testing

CI/CD

## Documentation

No. Covered by existing documentation.

dthaler commented:
PR description says:
> **Testing**
> CI/CD

But the issue linked doesn't say it's a bug.  Seems to me like the testing needed is what #1911 covers.

Alan-Jowett commented:
> PR description says:
> 
> > **Testing**
> > CI/CD
> 
> But the issue linked doesn't it's causing any CI/CD failures now. Seems to me like the testing needed is what #1911 covers.

Agreed. I will add some tests to unit tests to hit the negative path where program data info changes.

Alan-Jowett commented:
> > PR description says:
> > > **Testing**
> > > CI/CD
> > 
> > 
> > But the issue linked doesn't it's causing any CI/CD failures now. Seems to me like the testing needed is what #1911 covers.
> 
> Agreed. I will add some tests to unit tests to hit the negative path where program data info changes.

Added negative tests and fixed a bug that was found.

dthaler commented on libs/execution_context/ebpf_program.c:
```diff
@@ -1478,15 +1477,16 @@ _ebpf_helper_id_to_index_compare(const void* lhs, const void* rhs)
  * @brief Compute the hash of the program info and compare it with the hash stored in the program. If the hash does not
- * match then the program was verified against the wrong program info.
+ * match then the program was verified against the wrong program info. If the hash is not present then store the hash
+ * in the program to be used for future verification.
```
What does "future verification" mean?  Another pass by the verifier?

Alan-Jowett replied:
Improved the text for clarity.

dthaler commented on libs/execution_context/ebpf_program.c:
```diff
@@ -1614,14 +1614,26 @@ _ebpf_program_verify_program_info_hash(_In_ const ebpf_program_t* program)
-        goto Exit;
-    }
+    if (program->parameters.program_info_hash_length == 0) {
+        // This is the first time the program info hash is being computed.
```
Why is this done on first attach rather than at verification time?

Alan-Jowett replied:
Add comments explaining why this is the correct behavior. 

An instance of an ebpf_program_t is bound to a specific program information provider, which is fixed when the object is initially created by the user process. This locks the program information provider to that specific program information provider version. The verifier then queries the program object to get the program information, which is essentially immutable (if it changes it won't rebind).

dthaler commented on tests/end_to_end/end_to_end.cpp:
```diff
@@ -2685,3 +2685,121 @@ TEST_CASE("test_ebpf_object_set_execution_type", "[end_to_end]")
+    // Test that we drop the packet and increment the map
+    xdp_md_t ctx0{packet0.data(), packet0.data() + packet0.size(), 0, TEST_IFINDEX};
+
+    // Try loading with out the extension loaded.
```
```suggestion
    // Try loading without the extension loaded.
```

Alan-Jowett replied:
As it exists today, the hash is computed prior to knowing which helpers functions the program uses. This forces any change to the program information to require a reverification.

dthaler commented on tests/end_to_end/end_to_end.cpp:
```diff
@@ -2685,3 +2685,121 @@ TEST_CASE("test_ebpf_object_set_execution_type", "[end_to_end]")
+        REQUIRE(hook_result != XDP_PASS);
+    }
+
+    // Reload the extension provider with changed helper function data
```
What about a test case that merely adds a new helper function (which of course is not called by the program)?

Alan-Jowett replied:
As it exists today, the hash is computed prior to knowing which helpers functions the program uses. This forces any change to the program information to require a reverification.

dv-msft replied:
We can file a new issue to address the (incremental) edge case raised by Dave.

dv-msft commented on libs/platform/user/ebpf_leak_detector.cpp:
```diff
@@ -54,6 +54,7 @@ _ebpf_leak_detector::dump_leaks()
         }
         std::cout << std::endl;
     }
+    ebpf_assert(_allocations.empty());
```
This can be removed as it's now being done just below (pulled in from a different PR that got merged).

dthaler commented on libs/execution_context/ebpf_program.c:
```diff
@@ -1478,15 +1477,26 @@ _ebpf_helper_id_to_index_compare(const void* lhs, const void* rhs)
+ * in the program so it can be compared when the program information provider reattaches.
+ *
+ * Notes on why this works:
+ * 1) The user application creates a ebpf_program_t object and set the program type.
```
grammar nit:
```suggestion
 * 1) The user application creates an ebpf_program_t object and sets the program type.
```

dthaler commented on libs/execution_context/ebpf_program.c:
```diff
@@ -1478,15 +1477,26 @@ _ebpf_helper_id_to_index_compare(const void* lhs, const void* rhs)
+ * Notes on why this works:
+ * 1) The user application creates a ebpf_program_t object and set the program type.
+ * 2) During initialization, the program binds to the program information provider.
+ * 3) During the attach callback, the program information is hashed and stored.
```
One can load a program without attaching, it and verification has to happen on load.  So why is 3 before 4?

Alan-Jowett replied:
Need to clarify what attach means in this context. This is the NMR attach to the program information provider (not the BPF attach point).

dthaler commented on libs/execution_context/ebpf_program.c:
```diff
@@ -1478,15 +1477,26 @@ _ebpf_helper_id_to_index_compare(const void* lhs, const void* rhs)
+ * 2) During initialization, the program binds to the program information provider.
+ * 3) During the attach callback, the program information is hashed and stored.
+ * 4) The verifier then queries the program information from the ebpf_program_t object and uses it to verify the program
+ * safety. 5) If the program information provider is reattached, the program information is hashed and compared with the
```
```suggestion
 * safety. 
 * 5) If the program information provider is reattached, the program information is hashed and compared with the
```

dthaler commented on tests/end_to_end/end_to_end.cpp:
```diff
@@ -2685,3 +2685,121 @@ TEST_CASE("test_ebpf_object_set_execution_type", "[end_to_end]")
+    // Create a 0-byte UDP packet.
+    auto packet0 = prepare_udp_packet(0, ETHERNET_TYPE_IPV4);
+
+    // Test that we drop the packet and increment the map
```
```suggestion
    // Test that we drop the packet and increment the map.
```
consistency nit

dthaler commented on tests/end_to_end/end_to_end.cpp:
```diff
@@ -2685,3 +2685,121 @@ TEST_CASE("test_ebpf_object_set_execution_type", "[end_to_end]")
+        REQUIRE(hook_result != XDP_PASS);
+    }
+
+    // Reload the extension provider with changed helper function data
```
```suggestion
    // Reload the extension provider with changed helper function data.
```

dthaler reviewed (commented):
Holding off on clicking approve because I don't understand the comment.  You should be able to load a program without attaching it yet.  How does that work?

dthaler commented on libs/execution_context/ebpf_program.c:
```diff
@@ -1478,15 +1477,27 @@ _ebpf_helper_id_to_index_compare(const void* lhs, const void* rhs)
+ * Notes on why this works:
+ * 1) The user application creates an ebpf_program_t object and sets the program type.
+ * 2) During initialization, the program binds to the program information provider.
+ * 3) During the attach callback, the program information is hashed and stored.
```
```suggestion
 * 3) During the NMR attach callback, the program information is hashed and stored.
```
